### ISSUE TITLE
Bug: WASM not training?

### ISSUE THREAD

**Issue by kwhitley** [DanRuta/jsNet#47]
No change to structure except enabling/disabling the WASM...

Here's the (slow) JS version training:
![image](https://user-images.githubusercontent.com/865416/37796520-83d2b3da-2de5-11e8-9b2f-749fefbef79f.png)

And now the WASM version:
![image](https://user-images.githubusercontent.com/865416/37796591-a47b9c46-2de5-11e8-9356-786ea2a0ca11.png)

**Note**
The WASM does not appear to be training/updating based on the cost/loss output...

Ideas?

**Comment by DanRuta** [comment]
Hmm.. Did you use Conv/Pool Layers, or just FC?

**Comment by kwhitley** [comment]
Just FC (this is still previous network).  Did not implement the new `InputLayer` if that matters... just a bunch of FC layers.

This is using relu activation, crossEntropy and a softmax output (although I left the code as-is from before updating to 3.3.1, so perhaps I have to declare softmax output explicitly?

**Comment by kwhitley** [comment]
So just updated to InputLayer & OutputLayer... says OutputLayer is not a constructor, and does not appear to be exported/exposed from the webassembly() output (console.log shows it as undefined, where as InputLayer is showing correctly as a class extending FCLayer)

Am I missing something?

**Comment by DanRuta** [comment]
You may have an out of date version, OutputLayer came right after InputLayer. But both extend FCLayer, so I don't think that is the matter. Did you use a volume for the input, in the end, or 1D array?

**Comment by kwhitley** [comment]
Still 1D array... I wanted to get the WASM version working identically first before I started down the Conv/3D path

**Comment by kwhitley** [comment]
Ok, updated to 3.3.2 and the OutputLayer issue goes away... but the network still doesn't appear to train like when I switch it back to js()

**Comment by DanRuta** [comment]
Strange. I can try stepping through the values to catch discrepancies if you give me your config and a data sample.

**Comment by kwhitley** [comment]
Seem to have narrowed it down... the failure to learn appears to be introduced when using RMSProp with the webassembly version

```js
let net = window.net = new Network({
    Module,
    activation: 'relu',
    cost: 'crossEntropy',
    updateFn: 'RMSProp', // commenting this out shows the error in fact declining
    layers: [
      new InputLayer(inputLength),
      new FCLayer(round(inputLength * 0.5)),
      new FCLayer(round(inputLength * 0.3)),
      new FCLayer(round(inputLength * 0.2)),
      new FCLayer(round(inputLength * 0.1)),
      new OutputLayer(outputLength, { activation: 'sigmoid', softmax: true })
    ]
  })
```

**Comment by DanRuta** [comment]
Ah, nice one. I'll have a play with it. I'll get back to you on that.

**Comment by kwhitley** [comment]
From a little trial and error, the optimizers appear to fall into the following categories (with the WASM version at least)

###Working
- vanillasgd
- adagrad

###Not Working
- momentum
- RMSProp (or RMS_Prop, as I see it referenced once in the docs)
- adam
- adadelta
- gain

**Comment by kwhitley** [comment]
That said, I'm using adagrad at the moment and it's working... just figured RMS would be faster 🤷‍♂️

**Comment by DanRuta** [comment]
Turns out there was a bug in assigning default learning rate values to WebAssembly (was checking for undefined, but should have checked 0, because C++), so the learning rate was 0 unless defined. Explains why tests didn't catch it. Not too sure about adadelta/gain, as they seemed to work for me. I'll keep digging.

In reference to the RMSProp/RMS_Prop comment, the string configs are case/space/underscore insensitive.

**Comment by kwhitley** [comment]
odd... regarding the learning rate - pretty sure I've had learning rate defined the entire time.  I'll keep poking around though!  I will say, the speed boost makes iterating sooooo much better 🥇 

Also, random, but regarding dropout - might want to check with other libs on this, but you define it opposite to what I'd assume:

A dropout of 1 to me would mean 100% dropped connections, as a 0.3 would imply a 30% dropout rate/dropped connections.  I think you define it as the inverse (0.3 means *keep* 30% of the connections)

**Comment by DanRuta** [comment]
I see. I've mostly been following the cs231n course for learning things (http://cs231n.github.io/neural-networks-2/), and they had been referring to dropout using probability of keeping. 

I was on the fence about this one, and it did indeed used to be the other way around, but then I saw Tensorflow also has it as probability of keeping (http://cs231n.github.io/neural-networks-2/), and changed it.

**Comment by kwhitley** [comment]
hmmm, well if TensorFlow does it that way, prob best to keep it as-is.  Seems counterintuitive, but most examples people will be following currently will be TF ones.

Also, I *love* the confusion matrix printout... so nice to tack that at the callback of the train session to compare training/test!  Having now tested virtually every major ML package on NPM (ConvNet, Synaptic, Neataptic, Nen, DNN, Deeplearn.js, etc) - I'm pretty sure you are on track to have the most robust one of all, once/if you add some of the competing architectures (LSTM, RNN, GRU, NARX, etc)

**Comment by DanRuta** [comment]
I agree with the counter-intuitiveness, but doing it my own way would probably not be a good idea, haha.

Thank you, I do hope to keep extending this as much as I can. Currently working on a GPU version, though first. Feed-forward FC layers are done so far, but GLSL and WebAssembly contexts are not making it easy for me 😥.

Once that's done, I hope to (learn and) add more things.
